src/pymor/analyticalproblems/expressions.py

```diff
@@ -0,0 +1,307 @@
+
+    def numpy_expr(self):
+        first_ind = second_ind = ''
+        if len(self.first.shape) > len(self.second.shape):
+            second_ind = (['...']
+                          + ['newaxis'] * (len(self.first.shape) - len(self.second.shape))
+                          + [':'] * len(self.second.shape))
+            second_ind = '[' + ','.join(second_ind) + ']'
+        if len(self.first.shape) < len(self.second.shape):
+            first_ind = (['...']
+                         + ['newaxis'] * (len(self.second.shape) - len(self.first.shape))
+                         + [':'] * len(self.first.shape))
+            first_ind = '[' + ','.join(first_ind) + ']'
+
+        return f'({self.first.numpy_expr()}{first_ind} {self.numpy_symbol} {self.second.numpy_expr()}{second_ind})'
+
+    def __str__(self):
+        return f'({self.first} {self.numpy_symbol} {self.second})'
+
+
+class Neg(Expression):
+
+    def __init__(self, operand):
+        self.operand = operand
+        self.shape = operand.shape
+
+    def numpy_expr(self):
+        return f'(- {self.operand.numpy_expr()})'
+
+    def __str__(self):
+        return f'(- {self.operand})'
+
+
+class Indexed(Expression):
+
+    def __init__(self, base, index):
+        self.base, self.index = base, index
+        self.shape = base.shape[(1 if isinstance(index, Number) else len(index)):]
+
+    def numpy_expr(self):
+        index = [repr(self.index)] if isinstance(self.index, Number) else [repr(i) for i in self.index]
+        index = ['...'] + index
+        return f'{self.base.numpy_expr()}[{",".join(index)}]'
+
+    def __str__(self):
+        return f'{self.base}[{",".join(self.index)}]'
+
+
+class X(Expression):
+
+    def __init__(self, dim_domain):
+        self.dim_domain = dim_domain
+        self.shape = (dim_domain,)
+
+    def numpy_expr(self):
+        return 'x'
+
+    def __str__(self):
+        return 'x'
+
+
+class UnaryFunctionCall(Expression):
+
+    numpy_symbol = None
+
+    def __init__(self, arg):
+        self.arg = arg
+        self.shape = arg.shape
+
+    def numpy_expr(self):
+        return f'{self.numpy_symbol}({self.arg.numpy_expr()})'
+
+    def __str__(self):
+        return f'{self.numpy_symbol}({self.arg})'
+
+
+class UnaryReductionCall(Expression):
+
+    def __init__(self, arg):
+        self.arg = arg
+        self.shape = ()
+
+    def numpy_expr(self):
+        return f'{self.numpy_symbol}({self.arg.numpy_expr()}.reshape(input_shape + (-1,)), axis=-1)'
+
+    def __str__(self):
+        return f'{self.numpy_symbol}({self.arg})'
+
+
+def _convert_to_expression(obj):
+    if isinstance(obj, Expression):
+        return obj
+    elif isinstance(obj, Number):
+        return Constant(obj)
+    else:
+        obj = np.array(obj)
+        if obj.dtype == object:
+            raise NotImplementedError
+        return Constant(obj)
+
+
+def _broadcastable(first, second):
+    return all(f == s or f == 1 or s == 1 for f, s in zip(first[::-1], second[::-1]))
```

Comment:
I feel it's surprising that this assumes shapes as input rather than arrays.

Reply:
I have renamed the function.